### ISSUE TITLE
invalid memory read in jas_matrix_asl (jas_seq.c)

### ISSUE THREAD

**Issue by asarubbo** [jasper-software/jasper#88]
On 1.900.27:

```
warning: trailing garbage in marker segment (56 bytes)
warning: trailing garbage in marker segment (41 bytes)
warning: trailing garbage in marker segment (44 bytes)
warning: trailing garbage in marker segment (69 bytes)
warning: trailing garbage in marker segment (69 bytes)
warning: trailing garbage in marker segment (78 bytes)
warning: trailing garbage in marker segment (69 bytes)
warning: trailing garbage in marker segment (42 bytes)
warning: not enough tile data (748 bytes)
warning: bad segmentation symbol
warning: bad segmentation symbol
warning: bad segmentation symbol
warning: forcing negative ROI shift to zero (bitstream is probably corrupt)
ASAN:DEADLYSIGNAL
=================================================================
==26941==ERROR: AddressSanitizer: SEGV on unknown address 0x62c80000a400 (pc 0x7f28c74e48ee bp 0x7ffcececdb70 sp 0x7ffcececdaf0 T0)
==26941==The signal is caused by a READ memory access.
    #0 0x7f28c74e48ed in jas_matrix_asl /tmp/portage/media-libs/jasper-1.900.27/work/jasper-1.900.27/src/libjasper/base/jas_seq.c:376:11
    #1 0x7f28c7545f0e in jpc_dec_tiledecode /tmp/portage/media-libs/jasper-1.900.27/work/jasper-1.900.27/src/libjasper/jpc/jpc_dec.c:1107:6
    #2 0x7f28c7536cdf in jpc_dec_process_sod /tmp/portage/media-libs/jasper-1.900.27/work/jasper-1.900.27/src/libjasper/jpc/jpc_dec.c:658:7
    #3 0x7f28c75406b3 in jpc_dec_decode /tmp/portage/media-libs/jasper-1.900.27/work/jasper-1.900.27/src/libjasper/jpc/jpc_dec.c:425:10
    #4 0x7f28c75406b3 in jpc_decode /tmp/portage/media-libs/jasper-1.900.27/work/jasper-1.900.27/src/libjasper/jpc/jpc_dec.c:262
    #5 0x7f28c74a2b84 in jas_image_decode /tmp/portage/media-libs/jasper-1.900.27/work/jasper-1.900.27/src/libjasper/base/jas_image.c:444:16
    #6 0x509eed in main /tmp/portage/media-libs/jasper-1.900.27/work/jasper-1.900.27/src/appl/imginfo.c:219:16
    #7 0x7f28c65aa61f in __libc_start_main /var/tmp/portage/sys-libs/glibc-2.22-r4/work/glibc-2.22/csu/libc-start.c:289
    #8 0x419978 in _init (/usr/bin/imginfo+0x419978)

AddressSanitizer can not provide additional info.
SUMMARY: AddressSanitizer: SEGV /tmp/portage/media-libs/jasper-1.900.27/work/jasper-1.900.27/src/libjasper/base/jas_seq.c:376:11 in jas_matrix_asl
==26941==ABORTING
```

Testcase: https://github.com/asarubbo/poc/blob/master/00053-jasper-invalidread-jas_matrix_asl

**Comment by adarshdinesh** [comment]
mov rax, [rbp+var_10]
mov rax, [rax] // invalid memory access

https://github.com/mdadams/jasper/blob/master/src/libjasper/base/jas_seq.c#L354
Here pointer 'data' is corrupted and while trying to access it's value application crashes.

while ( v4 > 0 )
{
  v5 = *(_QWORD *)(a1 + 48);
  v7 = v6;
  while ( v5 > 0 )
  {
    *v7 = jas_fast32_asl(*v7, a2); // here pointer v7 is not validated
    --v5;
    ++v7
  }
}

**Comment by setharnold** [comment]
Agostino, did this get a CVE number yet?

Thanks

**Comment by adarshdinesh** [comment]
No

**Comment by adarshdinesh** [comment]
Am I eligible for one ?

**Comment by asarubbo** [comment]
Seth, it didn't. I will do it asap.

**Comment by thoger** [comment]
CVE-2017-5505
https://blogs.gentoo.org/ago/2017/01/16/jasper-invalid-memory-read-in-jas_matrix_asl-jas_seq-c/

**Comment by thoger** [comment]
d42b2388 (included in 1.900.25) is the first commit that crashes on this reproducer.  I've not seen crash with an earlier versions.  That commit either introduced or uncovered this problem.

**Comment by MaxKellermann** [comment]
Since this project has been mostly dead for several years, we created a fork which aims to fix all vulnerabilities (of which there are many).
This bug will be fixed by https://github.com/jasper-maint/jasper/pull/38 (merge pending)
